rasterio/_warp.pyx

```diff
@@ -348,150 +352,138 @@ def _reproject(
-        #                                    _gdal.GDALGenImgProjTransform, 
-        #                                    hTransformArg, 
-        #                                    maxerror )
-        #    psWOptions.pfnTransformer = _gdal.GDALApproxTransform
-        #else:
-        psWOptions.pfnTransformer = _gdal.GDALGenImgProjTransform
-        psWOptions.pTransformerArg = hTransformArg
-        psWOptions.hSrcDS = hdsin
-        psWOptions.hDstDS = hdsout
-        psWOptions.nBandCount = src_count
-        psWOptions.panSrcBands = <int *>_gdal.CPLMalloc(src_count*sizeof(int))
-        psWOptions.panDstBands = <int *>_gdal.CPLMalloc(src_count*sizeof(int))
-        if isinstance(source, tuple):
-            psWOptions.panSrcBands[0] = source.bidx
-        else:
-            for i in range(src_count):
-                psWOptions.panSrcBands[i] = i+1
-        if isinstance(destination, tuple):
-            psWOptions.panDstBands[0] = destination.bidx
-        else:
-            for i in range(src_count):
-                psWOptions.panDstBands[i] = i+1
-        log.debug("Set transformer options")
-
-        # TODO: alpha band.
+            warp_extras, "UNIFIED_SRC_NODATA", "YES")
+
+
+    if dst_nodata is not None and not _io.in_dtype_range(
+            dst_nodata, destination.dtype):
+        raise ValueError("dst_nodata must be in valid range for "
+                         "destination dtype")
+
+    psWOptions.padfDstNoDataReal = <double*>_gdal.CPLMalloc(src_count * sizeof(double))
+    psWOptions.padfDstNoDataImag = <double*>_gdal.CPLMalloc(src_count * sizeof(double))
+    for i in range(src_count):
+        psWOptions.padfDstNoDataReal[i] = dst_nodata
+        psWOptions.padfDstNoDataImag[i] = 0.0
+    warp_extras = _gdal.CSLSetNameValue(
+        warp_extras, "INIT_DEST", "NO_DATA")
+
+    # Important: set back into struct or values set above are lost
+    # This is because CSLSetNameValue returns a new list each time
+    psWOptions.papszWarpOptions = warp_extras
+
+    psWOptions.pfnTransformer = _gdal.GDALGenImgProjTransform
+    psWOptions.pTransformerArg = hTransformArg
+    psWOptions.hSrcDS = hdsin
+    psWOptions.hDstDS = hdsout
+    psWOptions.nBandCount = src_count
+    psWOptions.panSrcBands = <int *>_gdal.CPLMalloc(src_count*sizeof(int))
+    psWOptions.panDstBands = <int *>_gdal.CPLMalloc(src_count*sizeof(int))
+    if isinstance(source, tuple):
+        psWOptions.panSrcBands[0] = source.bidx
+    else:
+        for i in range(src_count):
+            psWOptions.panSrcBands[i] = i+1
+    if isinstance(destination, tuple):
+        psWOptions.panDstBands[0] = destination.bidx
+    else:
+        for i in range(src_count):
+            psWOptions.panDstBands[i] = i+1
+    log.debug("Set transformer options")
 
-        eErr = oWarper.Initialize(psWOptions)
-        if eErr == 0:
+    # TODO: alpha band.
 
-            log.debug("Destination shape: %r", destination.shape)
+    if oWarper.Initialize(psWOptions):
```

Comment:
@sgillies how about `if oWarper.Initialize(psWOptions) != 0:`?

Otherwise, without a priori knowledge that this will return 0 for good and >0 for bad, this is non-intuitive (sort of implies if it initialized, then raise an error).